### ISSUE TITLE
Can you give me some detail about how you mask out unexpected paddings.

### ISSUE THREAD

**Issue by menglin0320** [JianshuZhang/WAP#9]
I implemented it but I really worry if it really works and it's implemented efficiently, can I have some hint.

**Comment by JianshuZhang** [comment]
you can mask out them only when computing attention probabilities

**Comment by menglin0320** [comment]
This one may be specific，on paper Image to mark generation with coarse-to-fine attention, apaper was published prior to your work, part of the model is a lstm used for localization. your model didn't have it, so you think it doesn't help?
I'm saying the part named as row encoder.

**Comment by JianshuZhang** [comment]
Actually we are parallel work，when we submit our WAP paper to Journal, WYSIWYG has not been published. However, we also tried the row encoder introduced in WYSIWYG，it didnot work on CROHME competition. But the CRNN framework may be better when hanwritten formula input is longer.

**Comment by menglin0320** [comment]
Thank you very much, you are such a generous researcher. Can I ask how you shrink the size of the mask to the right size correspond to the feature vector, I just did four max pool on the mask lol.
And I just want to make sure I'm right,
here is what I'm doing:
alpha_t = tf.exp(e)
alpha_t = tf.multiply(flatted_mask,alpha_t)
alpha_t = alpha_t/ tf.expand_dims(tf.reduce_sum(alpha_t,axis = -1),1)
is this what you mean only when computing attention probabilities.

**Comment by JianshuZhang** [comment]
Yes, your mask is performed correctly.

Compared with those popular research problems, fewer researchers pay attention to HMER. We hope to have more people's attention on HMER by sharing our methodology!

**Comment by menglin0320** [comment]
Hmmm, the traditional methods for OCR does segmentation first, I know it doesn't work well for complex math formulas but I didn't see any paper segmenting handwritten characters with deep nets. Do you know why this is the case? Are they just satisfied with traditional methods.

**Comment by menglin0320** [comment]
offline.

**Comment by JianshuZhang** [comment]
At first, we tried using RCNN to segment handwritten characters and it also worked well. But we finally employed attention to segment handwritten characters as it can be well jointly optimized with recognition.

**Comment by menglin0320** [comment]
Hmmm, by rcnn you mean faster rcnn? So you got a data set of math formulas with bounding boxes? Does Crohme provide that? I guess we can infer boxes from strokes but is the dataset big enough? 
I guess you did that for an internship so though I will ask what I wonder but you can refuse to answer. 
By good how good was that, and how much data did you have? 
I'm doing a task very similar to what you did, I'd be very thankful if I can get advise from you.

**Comment by JianshuZhang** [comment]
Yes, faster rcnn, get BB by using stroke information. We only use official training dataset (8835 expressions) and I didn't try more data further.
The 'good' means the idea works, but still worse than attention, maybe due to short of training data. As for suggestion, I prefer trying attention based encoder-decoder model, which can be trained jointly and conveniently, and is likely to achieve better performance.

**Comment by menglin0320** [comment]
Hmm and for weight noise did you use gaussian with constant varience or if this is the case I feel doing it in batch norm may improve the performance a little bit?
![image](https://user-images.githubusercontent.com/14263971/41237934-da4135b8-6d62-11e8-85cb-b5a7bc8b4a92.png)

**Comment by menglin0320** [comment]
And kine of curious how you find this regularizor.

**Comment by JianshuZhang** [comment]
Yes, the regularizor shows great improvement in Recurrent Neural Networks, which I have been researched in for a long time.

**Comment by menglin0320** [comment]
So you used the adaptive gaussian noise?
And the way to do that is initialize the sigma and then update the sigma by first calculating the variance by the equations shown above.

**Comment by JianshuZhang** [comment]
yes

发自我的 iPhone

**Comment by menglin0320** [comment]
hmm, you only apply the regularizer on lstm or you applied it on all weights

**Comment by JianshuZhang** [comment]
all weights



xysszjs@mail.ustc.edu.cn
 
From: menglin0320
Date: 2018-06-22 22:09
To: JianshuZhang/WAP
CC: Jianshu Zhang; Comment
Subject: Re: [JianshuZhang/WAP] Can you give me some detail about how you mask out unexpected paddings. (#9)
hmm, you only apply the regularizer on lstm or you applied it on all weights

**Comment by menglin0320** [comment]
So we are learning mean and variance of weight tensors instead of simply learning numbers for weights? I still feel it is not intuitive, did you try to only apply it on rnns because the original paper only did experiments on rnns.

My understanding is each weight tensor is Gaussian with diagonal variance so we can say each weight is sampled from G(u_i,sigma_i), Not rigorously speaking, on the paper it correspond to the posterior distribution.

u and sigma without _i are parameters of a prior distribution and we assume the prior distribution is simply a Gaussian distribution, and they are calculated as (or rigorously speaking, their optimal value is)
![image](https://user-images.githubusercontent.com/14263971/42045040-8c0b2c28-7ac8-11e8-8d93-1a51fd8414aa.png)
and the regularization term is (where u is same as hat{u} and sigma is same as hat{sigma}):
![image](https://user-images.githubusercontent.com/14263971/42044933-47909920-7ac8-11e8-81c4-7fc49d2da886.png)
we don't have to care about 
![image](https://user-images.githubusercontent.com/14263971/42045163-d23c98bc-7ac8-11e8-9876-ffbd8b547d34.png)
because deeplearning packages calculates this for you.
![image](https://user-images.githubusercontent.com/14263971/42045986-d9e70032-7aca-11e8-94e2-727ea75676e8.png)
I guess that W is size of one weight tensor instead of number of parameters on all weight tensor
. (In other words we apply the regularizer on each weight tensor separately) But on the paper's analysis W should be number of all weights on the model because the Gaussian prior is defined for whole model. intuitively I feel one Gaussian prior for whole network won't work. But I have to make sure, may be I'm wrong again. Also I'm thinking if it is worth trying to give each weight matrix(2d) a independent prior Gaussian distribution.

**Comment by menglin0320** [comment]
And did you use any reg parameter or this one doesn't need reg parameter with the theoretic analysis.

**Comment by JianshuZhang** [comment]
Yes, the weightnoise is originally devised for RNN, it does greatly improved the performance when the encoder and the decoder are both RNNs (you can refer to my repository TAP). But when the encoder is CNN, the gain is not such good and I am not sure whether it is better to apply weightnoise only at RNN. When I employ densenet as the encoder, I didn't choose to employ weightnoise any more as it takes too much time and no improvement. So, I prefer to choose DenseNet as the encoder without weightnoise as the weightnoise works perfectly on RNN networks.
As for the details of our implemented weightnoise, you can refer to the function 'apply_adative_noise' in nmt_weightnoise.py. Hope it will help you.



xysszjs@mail.ustc.edu.cn
 
From: menglin0320
Date: 2018-06-28 23:57
To: JianshuZhang/WAP
CC: Jianshu Zhang; Comment
Subject: Re: [JianshuZhang/WAP] Can you give me some detail about how you mask out unexpected paddings. (#9)
So we are learning mean and variance of weight matrices instead of simply learning numbers for weights? I still feel it is not intuitive, did you try to only apply it on rnns because the original paper only did experiments on rnns.
My understanding is each weight matrix is Gaussian with diagonal variance so we can say each weight is sampled from G(u_i,sigma_i), Not rigorously speaking, on the paper it correspond to the posterior distribution.
u and sigma without _i are parameters of a prior distribution and we assume it is simply a Gaussian distribution, and they are calculated as (or rigorously speaking, their optimal value is)

and the regularization term is (where u is same as hat{u} and sigma is same as hat{sigma}):

we don't have to care about

because deeplearning packages calculates this for you.
I guess that W is size of one weight matrix instead of number of parameters on all weight matrix. (In other words we apply the regularizer on each weight matrix separately) But on the paper's analysis W should be number of all weights on the model, intuitively I feel it won't work. There is a gap that is not trivial so I want to make sure.

**Comment by menglin0320** [comment]
I guess 
p_ls2 = theano.shared((numpy.zeros_like(p.get_value()) +
                               numpy.log(init_sigma) * 2. / log_sigma_scale
                               ).astype(dtype=numpy.float32) ,name='%s_sigma' % k) # log_(sigma^2)
p_s2 = tensor.exp(p_ls2 * log_sigma_scale)
is just a smart way to calculate sigmas while avoiding numerical error

So yes you are using a global prior which is Gaussian and you did it by calculating the gradients, I feel with deep learning architectures it's just more convient to add regularization term Lc(alpha,beta) to the loss. I'll try to do it with only lstm though it doesn't have any theoretic analysis.
Just to make sure that when you use adaptive weight noise, each time you do forward propogation you have to sample from a Gaussian distribution?

**Comment by JianshuZhang** [comment]
I guess your 'each time' means each batch during training, if so, Yes.



xysszjs@mail.ustc.edu.cn
 
From: menglin0320
Date: 2018-06-29 22:21
To: JianshuZhang/WAP
CC: Jianshu Zhang; Comment
Subject: Re: [JianshuZhang/WAP] Can you give me some detail about how you mask out unexpected paddings. (#9)
I guess
p_ls2 = theano.shared((numpy.zeros_like(p.get_value()) +
numpy.log(init_sigma) * 2. / log_sigma_scale
).astype(dtype=numpy.float32) ,name='%s_sigma' % k) # log_(sigma^2)
p_s2 = tensor.exp(p_ls2 * log_sigma_scale)
is just a smart way to calculate sigmas while avoiding numerical error
So yes you are using a global prior which is Gaussian and you did it by calculating the gradients, I feel with deep learning architectures it's just more convient to add regularization term Lc(alpha,beta) to the loss. I'll try to do it with only lstm though it doesn't have any theoretic analysis.
Just to make sure that when you use adaptive weight noise, each time you do forward propogation you have to sample from a Gaussian distribution?

**Comment by menglin0320** [comment]
Thanks

**Comment by menglin0320** [comment]
Hmmm, did you do anything special on batch norm, otherwise I feel calculated mean and varience will be influenced by padded empties.

**Comment by JianshuZhang** [comment]
I once used mask in BN to remove the influence of padded empties when calculating mean or var, but there is no difference in performance. So, you may not need to worry about this.

发自我的 iPhone
